here-naksha-lib-handlers/src/main/java/com/here/naksha/lib/handlers/SourceIdHandler.java

```diff
@@ -114,14 +125,66 @@ private Optional<String> getSourceIdFromFeature(NakshaProperties properties) {
     }
   }
 
-  public static Optional<ITagQuery> mapIntoTagOperation(PQuery propertyOperation) {
+  private static boolean isFullyConvertedToITagQuery(IPropertyQuery propertyQuery) {
+    return (propertyQuery instanceof PQuery)
+        || (propertyQuery instanceof PNot)
+        || (propertyQuery instanceof POr)
+        || ((propertyQuery instanceof PAnd pAnd) && (pAnd.isEmpty()));
+  }
 
-    if (sourceIdTransformationCapable(propertyOperation) && operationTypeAllowed(propertyOperation)) {
-      final TagExists tagQuery = new TagExists(TAG_PREFIX + propertyOperation.getValue());
-      return Optional.of(tagQuery);
+  private static Optional<ITagQuery> transformPropertyOperation(IPropertyQuery propertyOperation) {
+    if (propertyOperation instanceof PAnd pAnd) {
+      final TagAnd tagAnd = new TagAnd();
+      // List of successfully transformed property queries to be removed at the end, so as not to disrupt the loop
+      final List<IPropertyQuery> toRemove = new ArrayList<>();
+      final int size = pAnd.size();
+      for (int i = 0; i < size; i++) {
+        final IPropertyQuery propertyComponent = pAnd.get(i);
+        final Optional<ITagQuery> tagComponent = transformPropertyOperation(propertyComponent);
+        if (tagComponent.isPresent()) {
+          tagAnd.add(tagComponent.get());
+          if (isFullyConvertedToITagQuery(propertyComponent)) {
+            toRemove.add(propertyComponent);
+          }
+          // Unwrap the query if it is an AND clause with only 1 remaining sub-clause
+          else if ((propertyComponent instanceof PAnd canBeSimplified) && (canBeSimplified.size() == 1)) {
+            pAnd.set(i, canBeSimplified.get(0));
+          }
+        }
+      }
+      pAnd.removeAll(toRemove);
+      if (tagAnd.isEmpty()) {
+        return Optional.empty();
+      } else if (tagAnd.size() == 1) {
```

Comment:
I wonder if these simplifications are really needed - they will happen only if the client will specify property AND with single tiem, right?

Reply:
It's for cases where a given property query PAnd is not fully transformed to tag query. Let's take the simplest example, a read features request with only this property query, no tag or other query:

`(p.foo=bar && p.@ns:com:here:mom:meta.sourceId=abc)`

The second clause can be converted into if tag `xyz_source_id_abc` exists, but the first stays the same. As we resolve this `PAnd`, we assume that it can possibly be converted fully into tag query, so the result is also a `TagAnd`. But after conversion we have a `PAnd` with a single component (the 1st clause unchanged) and a TagAnd with also one single component (the 2nd clause transformed). 

Technically these kinds of query are allowed, but I'm not sure if the other handler components and ultimately the storages can resolve those correctly, considering these can be nested as well, like a PAnd containing only 1 PAnd containing only 1 PAnd... So we have 2 separate simplifications there, one for PAnd and one for TagAnd.